### PR TITLE
Share recipie option

### LINKED ISSUE #152: FEATURE:  Implement Recipe Sharing Feature
### Is there an existing issue raised for this?

- [X] I have searched the existing issues

### Describe the feature

I propose adding a Recipe Sharing Feature that allows users to share their favorite recipes directly on social media platforms such as Facebook, Twitter, and Pinterest.

### Component

This feature will be added to the existing recipe detail page, specifically at the URI: /recipes/:id.

### It should be implemented because

->Enhance user engagement by encouraging users to showcase their favorite recipes.
->Increase visibility for the application as recipes are shared across social media platforms.
->Foster a sense of community, allowing users to connect over shared culinary interests.

### Would you like to work on this issue?

Yes

### Terms & Conditions

- [X] I agree to follow the Code of Conduct
- [X] I'm a GSSOC'24 Ext contributor

### PR DESCRIPTION
**Pull request by raj-adi00**
## What does this PR do?

This PR solves the issue #152  . Here i have added backend code so that when user clicks share button automatically number of shares for that recipe gets updated by one. After that i update number of share on client side by using UseState. I don't want to make unnecessary call to backend so i increase it by one. But when page is loaded again fresh data has updated number of shares. As share button is pressed I access current web url and add id of the post at last which becomes url of paticular recipe. It is copied in clipboard and user can share it. When web is hosted it will give correct url becuase it will have proper domain. I have also updated that page that displays individual Recipe. In that page state was passed from all Recepie page but if we try to access that using individual url then it will cause error as state was not passed at that time. So i updated code that when there is no state then it should request backend for content of that page using id of individual recipe. This will help in case if state is not passed or any user tries to access individual recipe using url that has been shared to him. I have also added feature that when user click share button he gets option to go to different social media like whatsapp ,facebook where he can share these thing.

If you don't want the last feature then i can simply copy to clipboard and show that to user that it has been copied. Please mention whether you want the last feature or not. 

i have installed react icons for generating icons of different social media.

Fixes # (issue)
Fixes #152 

Screenshot 2024-10-21 193812
Screenshot 2024-10-21 193832

## How should this be tested?
Click on the share button and see the changes. Run the copied url in another tab. 

I tested responsiveness of my component and shared the recipie. I checked database it was properly updated. I tested url in another chrome on the same device.
Please note currently it is copying localhost because i was running it on  my localhost if user will run it on hosted website he will be able to capture url of that website and it will work perfectly fine. I didn't hard code it i have used window.location so that i can capture actual website url and used routing to properly direct to any page so that if any user visits the url he can see that.

- Test 1
- Test 2

### Required

- [x ] Filled out the "How to test" section in this PR
- [x ] Read Contributing guide
- [x ] Self-reviewed my own code
- [x ] Checked for warnings, there are none
- [ x] Removed all `console.logs`
- [x ] The changes don't cause any responsiveness issues

Requst: Can you please add level 3 if you feel i have worked as per that

### CODE REVIEW

**Comment by raj-adi00** [comment]
Ya sure!!

**Comment by raj-adi00** [comment]
https://github.com/user-attachments/assets/5231bf00-984c-4393-8b4d-4ad312824ce6

At bottom there are icons of social media but it couldn't capture. it is something like this
image
Clicking on any icon direct you to that site

**Comment by raj-adi00** [comment]
I found that when we use that copied url then date shows NaN-NaN-NaN so i fixed it later by keeping it in useEffect now this issue is solved

**Comment by AlfiyaSiddique** [comment]
@raj-adi00 some improvement!
Just show the pop-up toast that the link has copied no need for that black div and also to show the entire link address in screen.
Just make the toast alert appear

**Comment by raj-adi00** [comment]
image
do you want something like this

**Comment by raj-adi00** [comment]
@AlfiyaSiddique is there any option for success toat. I did not seet it.

**Comment by AlfiyaSiddique** [comment]
There is a function toast.success()

**Comment by raj-adi00** [comment]
image
@AlfiyaSiddique  done

File: backend/Controllers/RecipeController.js [hunk 1]
```diff
@@ -129,10 +130,10 @@ const imageToGithub = async (fileImage, name, unique) => {
   }
 
   console.log('Config:', { owner, repo, branch }); // Debug log
-
+  console.log(fileImage)
   const base64Content = fileImage.split(';base64,').pop();
```
Comment (AlfiyaSiddique) [inline backend/Controllers/RecipeController.js:134]: @raj-adi00 remove this console.log

File: backend/server.js [hunk 2]
```diff
@@ -13,12 +13,13 @@ const port = process.env.PORT;
...
 app.use(express.json({ limit: "10mb" }));
-app.use(express.urlencoded({ limit: "10mb" }));
+app.use(express.urlencoded({ limit: "10mb", extended: true }));
 
 // CORS configuration
 const allowedOrigins = [
     "https://delightful-daifuku-a9f6ea.netlify.app",
     /https:\/\/deploy-preview-\d+--delightful-daifuku-a9f6ea\.netlify\.app/,
+    "http://localhost:5173"
 ];
```
Comment (AlfiyaSiddique) [inline backend/server.js:23]: @raj-adi00 Remove this localhost link

File: backend/server.js [hunk 3]
```diff
@@ -37,7 +38,7 @@ app.use(
     })
 );
 
-// app.use(cors({origin:"http://localhost:5173"})) // for local use
+app.use(cors({ origin: "http://localhost:5173" })) // for local use
```
Comment (AlfiyaSiddique) [inline backend/server.js:42]: Comment this line as it was before

File: frontend/src/Pages/OneRecipe.jsx [hunk 4]
```diff
@@ -1,68 +1,108 @@
...
+        console.error("Error fetching comments:", error);
+      }
+    };
+    newDate(new Date(recipe.date))
+    fetchComments();
+  }, [recipe]);
+
+
+
+
```
Comment (AlfiyaSiddique) [inline frontend/src/Pages/OneRecipe.jsx:72]: So many empty lines here remove them keep max 2 lines as empty

**Comment by AlfiyaSiddique** [comment]
@raj-adi00 please update your PR and some code improvement please fix it and will merge the PR

**Comment by raj-adi00** [comment]
@AlfiyaSiddique  done the changes you told. Please review

File: backend/routes/web.js [hunk 5]
```diff
@@ -7,9 +7,15 @@ import authenticateToken from "../middleware/auth.js";
...
+<<<<<<< HEAD
+router.get('/usernames', UserController.getAllUserName)
+router.get('/token', authenticateToken, UserController.verifyUserByToken)
+router.get("/recipes", RecipeController.allRecipe)
+=======
 router.get("/usernames", UserController.getAllUserName);
 router.get("/token", authenticateToken, UserController.verifyUserByToken);
 router.get("/recipes", RecipeController.allRecipe);
+>>>>>>> 2c3a6840f60182553c9345c7f4a9af5057c8f6a0
 // added route to get previous comments
```
Comment (AlfiyaSiddique) [inline backend/routes/web.js:19]: @raj-adi00 the above <<<< Head like part will give syntactical error please remove it. This is due to not resolving the conflicts properly

**Comment by raj-adi00** [comment]
@AlfiyaSiddique  done

**Review by AlfiyaSiddique — APPROVED** [review]

**Pull request merged**